### ISSUE TITLE
Migrate to webpack/rollup and ditch requirejs

### ISSUE THREAD

**Issue by saschanaz** [w3c/respec#1975]
I'm not sure what we should do for existing `text!` thing yet.

**Comment by saschanaz** [comment]
Hmm. @marcoscaceres Thoughts?

**Comment by saschanaz** [comment]
Tried this a bit and found we should first remove the `deps/text` things.

**Comment by marcoscaceres** [comment]
We should be able to use one of the CSS importer things for text, I think. Or are there actual text we import? 

Agree about the UI. We should be able to ditch dynamic imports.

**Comment by saschanaz** [comment]
A question: Do we want to keep our babel transpiler after webpack/rollup migration? Maybe for debugging, as sourcemaps are not perfect yet?

**Comment by sidvishnoi** [comment]
We need babel until Safari supports object spread assignment though.

**Comment by saschanaz** [comment]
webpack/rollup internally have babel so I think it will be okay.

**Comment by saschanaz** [comment]
#1981 made the webpack part but not really ditched requirejs.

**Comment by saschanaz** [comment]
Currently I have two options:

1. Migrate raw text imports to `fetch()` with a custom webpack plugin that converts `fetch()` into text importer call.
2. Implement a service worker that understands text import request and serves proper JavaScript module instead.

Option 1 needs some study because I don't know details to implement such plugins. Option 2 should be easier but 1) needs a way to detect whether a request is for text import or not 2) the worker should be injected separately to a document because the webpack-built version won't need it.

**Comment by marcoscaceres** [comment]
> Implement a service worker that understands text import request and serves proper JavaScript module instead.

Good idea - but that's not going to work, unfortunately :( The same origin restriction would mean that each spec would also need to locally host a copy of the service worker. 

If the text calls are mostly CSS... we could consider CSS-in-JS?

**Comment by saschanaz** [comment]
>The same origin restriction would mean that each spec would also need to locally host a copy of the service worker.

We can keep using webpack to bundle CSS. The suggested service worker is only for non-bundled development version.

**Comment by marcoscaceres** [comment]
Ah, ok. It still doesn't feel right that we would need two versions... from experience, the bundled version should still be debugguble in WebPack (just going to be a big fat file)... so I'm still hopeful we can do this without resorting to 2. 

I need to explore exactly what stuff is coming in with !text.

**Comment by saschanaz** [comment]
>V

**Comment by saschanaz** [comment]
BTW, we can also replace `require()` with native import from our tests if we remove any loader-specific syntax 👍

**Comment by marcoscaceres** [comment]
> They are mostly CSS, but personally it sounds not best... Hmm.

Agree, also makes me a bit 🤢. I think just have a play and experiment and see what works. I don't have a strong opinion here: just want to make sure ReSpec stays easy to set up and hack on (and hopefully we get the bundle size down a little bit more). 

> BTW, we can also replace require() with native import from our tests if we remove all loader-specific syntax 👍

That would be nice.

**Comment by saschanaz** [comment]
Chromium people just published an intent to implement HTML modules, so maybe one day we can do:

```html
<style id="style">
  .myCSSRule {}
</style>
<script type="module">
  export default style.textContent;
</script>
```

Still not perfect but this way at least IDEs can still syntax highlight and linters can still work 😆

**Comment by saschanaz** [comment]
We have too many blockers. Currently requirejs is only used for debugging, how about we add [a non-optimized mode](https://webpack.js.org/concepts/mode/) so that we can get a debugger-friendly build?

**Comment by marcoscaceres** [comment]
I'd be down with that - from another project I was working on, source mapping remains a disaster.

**Comment by saschanaz** [comment]
The main issue for me is that too many lines are optimized out so that breakpoints cannot be added consistently. What ware your problems there?

**Comment by marcoscaceres** [comment]
... but working with the non-optimized mode totally works, even if it means debugging at package with like 30,000+ lines.

**Comment by marcoscaceres** [comment]
> What were your problems there?

Just trying to set breakpoints and then stepping... it was completely erratic (jumps to random places, or can't actually set a breakpoint, etc).

**Comment by saschanaz** [comment]
Hmm... how about we add a non-optimized build option without actually removing requirejs and see how it works?

**Comment by marcoscaceres** [comment]
Sounds good to me :)

**Comment by marcoscaceres** [comment]
it seems totally fine to just have the non-optimized version for development, and hopefully be able to use webpack's development server too.

**Comment by saschanaz** [comment]
Note: the debug mode is added by #2149

**Comment by marcoscaceres** [comment]
Did a little more experimenting... I think we should do this:

 1. create `📂 profiles` folder.  
 1. profiles are called, for example, w3c.js. 
 1. create a dynamic Webpack config: 

```JS
const path = require("path");

module.exports = {
  mode: env.DEBUG ? "development" : "production",
  entry: `./profiles/${env.profile}.js`,
  output: {
    filename: "profile-${profile}.js",
    path: path.resolve(__dirname, "../builds"),
  },
};
```

## Profile structure

```JS
// New top level module...
import { runAll } from "../src/core/respec.js";

const plugins = [
  // plugins... for this profile
];

(async() => {
  respec.ui.show();
  await respec.load(plugins); // waits for DOMContentLoaded, runs base-runner runAll(), 
})();
```

We modify base-runner to import() plugins. 

### What about CSS, SVG?
For most modules, the CSS is small... let's just inline it or properly CSS-in-JS it.

**Comment by saschanaz** [comment]
Webpack can track the dependencies only in `import("...path...")` but not in `var path; import (path);` format, so probably we should do:

```js
const plugins: [
  import("./core/foo.js"),
  import("./core/bar.js"),
  // and more
];

(async() => {
  respec.ui.show();
  const plugs = await Promise.all(plugins);
  await respec.load(plugs);
})();
`
```

**Comment by marcoscaceres** [comment]
Another option is:

```
import {run as fooRun} from "./core/foo.js"
plugins = [
 fooRun,
]
```

**Comment by saschanaz** [comment]
That would double the lines and we have to manually name all the `run`s, so personally I don't think it's preferable.

**Comment by marcoscaceres** [comment]
Agree. Just mapping out all the options.

**Comment by marcoscaceres** [comment]
Kinda sad... feels like the end of an era. Oh well! 💃

**Comment by saschanaz** [comment]
It's not completely dead just yet, see #2552 😁
